integrations/gcp/gcp_core/helpers/ratelimiter/base.py

```diff
@@ -0,0 +1,178 @@
+                    f"No quota information found for '{self.service}:{self.quota_id}' in container '{container_id}'. "
+                    f"Default quota of {self._default_quota} will be used."
+                )
+                return self._default_quota
+
+            # Find the dimension with the least quota value
+            least_quota_info = min(
+                quota_infos, key=lambda info: int(info.details.value)
+            )
+            least_value = int(least_quota_info.details.value)
+
+            if len(quota_infos) > 1:
+                logger.info(
+                    f"Multiple quota dimensions found for '{self.service}:{self.quota_id}' in container '{container_id}'. "
+                    f"Selected the least value: {least_value} for further processing."
+                )
+
+            effective_quota_limit: int = int(
+                round(least_value * _PERCENTAGE_OF_QUOTA, 1)
+            )
+
+            logger.info(
+                f"Found quota information for '{self.service}:{self.quota_id}' in container '{container_id}' with value: {least_value} for locations: {least_quota_info.applicable_locations}."
+                f"The Integration will utilize {_PERCENTAGE_OF_QUOTA * 100}% of the quota, which equates to {effective_quota_limit} for rate limiting."
+            )
+
+            return effective_quota_limit
+
+        except GoogleAPICallError as e:
+            logger.warning(
+                f"Failed to retrieve quota from GCP for '{self.service}:{self.quota_id}' in container '{container_id}'. "
+                f"Default quota {self._default_quota} will be used. Error: {e}"
+            )
+            return self._default_quota
+
+        except Exception as e:
+            logger.error(
+                f"An internal server error occured while quota from GCP for '{self.service}:{self.quota_id}' in container '{container_id}'. "
+                f"Default quota {self._default_quota} will be used. Error: {e}"
+            )
+            return self._default_quota
+
+    @abstractmethod
+    def quota_name(self, container_id: str) -> str:
+        """Generate the fully qualified name for the quota resource."""
+        pass
+
+
+class GCPResourceRateLimiter(GCPResourceQuota):
+    """
+    GCPResourceRateLimiter manages rate limits based on GCP resource quotas.
+    It inherits from GCPResourceQuota and leverages GCP's quota information to dynamically determine and apply rate limits using the AsyncLimiter.
+    This class allows efficient control over the rate of API requests per container (e.g., projects, folders) based on the actual quota allocated by GCP.
+    """
+
+    time_period: float = _DEFAULT_RATE_LIMIT_TIME_PERIOD
+
+    @cache_coroutine_result()
+    async def register(self, container_id: str, *arg: Optional[Any]) -> AsyncLimiter:
+        quota = await self._get_quota(container_id, *arg)
+
+        effective_quota_limit: int = int(max(round(quota * _PERCENTAGE_OF_QUOTA, 1), 1))
```

Comment:
Do you need to apply the percentage again? You do that in the `effective_quota_limit` variable above

Reply:
No no, we don't